fenix/app/src/main/java/org/mozilla/fenix/shopping/state/ReviewQualityCheckAction.kt

```diff
@@ -9,4 +9,47 @@ import mozilla.components.lib.state.Action
 /**
  * Actions for review quality check feature.
  */
-sealed interface ReviewQualityCheckAction : Action
+sealed interface ReviewQualityCheckAction : Action {
+
+    /**
+     * Actions that are observed by middlewares.
+     */
+    sealed interface MiddlewareAction : ReviewQualityCheckAction
+
+    /**
+     * Actions that cause updates to state.
+     */
+    sealed interface UpdateAction : ReviewQualityCheckAction
+
+    /**
+     * Actions related to preferences.
+     */
+    sealed interface PreferencesMiddlewareAction : MiddlewareAction
```

Comment:
I'm on the fence about having these sealed interfaces for the actions to inherit from. There are undoubtedly going to be situations where one user input will trigger something that needs to be handled by multiple middlewares. 

Looking at `OptInAction` below, we'll be dispatching this when the user clicks on the preference toggle, but it's a `PreferencesMiddlewareAction` and is therefore only ever going to be consumed by the `PreferencesMiddleware`. Based on the current state of this, if there needed to be any additional side effects as a result of that user click (such as telemetry), we'll then have to fire additional actions for every specific middleware that needs to react. 

Or was your intent for Actions to implement any number of these super types, and `OptInAction` would eventually be a `PreferencesMiddlewareAction` and something like`TelemetryMiddlewareAction`?

Ultimately, I'm not sure there's much to gain having the overhead of these three super types vs just following the existing pattern of having the middlewares receive every action and simply ignoring the irrelevant ones. The middleware in this PR is already using a `when` for parsing out the relevant actions, there's now just an additional interstitial short-circuit check before it.

This all seems like a pattern on top of Redux, but I'm curious what @MatthewTighe would think.

Reply:
I left a comment in the middleware below that addresses my feelings on this a bit. In general, I agree with Noah. I think we risk introducing brittleness to this feature with coupling that is this specific.

Reply:
This is not too dissimilar to having `MessagingAction` sealed heirarchy in `AppAction`. The overall idea is to rely on the compiler and have all the cases covered and not use else. I've updated the actions so it will make more sense now. Let's get on a call if we want to discuss more.

Reply:
Would be happy to have this discussion in more depth offline, but thought I'd leave a comment since I'm already here.

The big difference I see with the `MessagingAction` is that it's shrinking specificity around a feature of a global Store. To me, it makes sense to add that level of granularity. That's mirrored in the fact that the state of the feature can be easily encapsulated in a `MessagingState` property on the greater `AppState`. There's also no tight coupling implied between `MessagingAction` and its middleware or reducer. Anything that interacts with `AppState` can potentially respond to `MessagingAction`s. 

My experience in prototyping to combine global Stores (and potentially move to a single `ScopedStore` model where feature stores could be temporarily combined) informed my opinion that in some cases we actually want to _broaden_ our specificity in order to create a more powerful tool. For example, I think we will need to remove the generic action bounds on pretty much every foundational type in `lib-state` in order to make those changes. We would definitely lose some compile-time exhaustiveness checking in our `when` expressions, but I think the overall benefits will be well worth it.

Reply:
To close the loop of this convo from my end, I think we acknowledged some of the "risks" of doing a state machine on top of Redux, so I'm fine with moving forward and seeing the final product of the code. I'm personally not 100% sold on the approach yet, but I want to be proven wrong and I trust you to do a good execution on this.

Reply:
sure thing! Happy to learn and iterate as we move forward 👍